command.go

```diff
@@ -1,253 +1,456 @@
+		namePath = append(namePath, cmd.parent.FullName())
+	}
 
-func (c CommandsByName) Len() int {
-	return len(c)
+	return strings.Join(append(namePath, cmd.Name), " ")
 }
 
-func (c CommandsByName) Less(i, j int) bool {
-	return lexicographicLess(c[i].Name, c[j].Name)
-}
+func (cmd *Command) Command(name string) *Command {
+	for _, subCmd := range cmd.Commands {
+		if subCmd.HasName(name) {
+			return subCmd
+		}
+	}
 
-func (c CommandsByName) Swap(i, j int) {
-	c[i], c[j] = c[j], c[i]
+	return nil
 }
 
-// FullName returns the full name of the command.
-// For subcommands this ensures that parent commands are part of the command path
-func (c *Command) FullName() string {
-	if c.commandNamePath == nil {
-		return c.Name
+func (cmd *Command) setupDefaults(arguments []string) {
+	if cmd.didSetupDefaults {
+		tracef("already did setup")
+		return
 	}
-	return strings.Join(c.commandNamePath, " ")
-}
 
-func (cmd *Command) Command(name string) *Command {
-	for _, c := range cmd.Commands {
-		if c.HasName(name) {
-			return c
-		}
+	cmd.didSetupDefaults = true
+
+	isRoot := cmd.parent == nil
+	tracef("isRoot? %[1]v", isRoot)
+
+	if cmd.ShellComplete == nil {
+		tracef("setting default ShellComplete")
+		cmd.ShellComplete = DefaultCompleteWithFlags(cmd)
 	}
 
-	return nil
-}
+	if cmd.Name == "" && isRoot {
+		tracef("setting cmd.Name from first arg basename")
+		cmd.Name = filepath.Base(arguments[0])
+	}
+
+	if cmd.Usage == "" && isRoot {
+		tracef("setting default Usage")
+		cmd.Usage = "A new cli application"
+	}
+
+	if cmd.Version == "" {
+		tracef("setting HideVersion=true due to empty Version")
+		cmd.HideVersion = true
+	}
+
+	if cmd.Action == nil {
+		tracef("setting default Action as help command action")
+		cmd.Action = helpCommandAction
+	}
 
-func (c *Command) setup(ctx *Context) {
-	if c.ShellComplete == nil {
-		c.ShellComplete = DefaultCompleteWithFlags(c)
+	if cmd.Reader == nil {
+		tracef("setting default Reader as os.Stdin")
+		cmd.Reader = os.Stdin
 	}
-	if c.Command(helpCommand.Name) == nil && !c.HideHelp {
-		if !c.HideHelpCommand {
-			helpCommand.HelpName = fmt.Sprintf("%s %s", c.HelpName, helpName)
-			c.Commands = append(c.Commands, helpCommand)
+
+	if cmd.Writer == nil {
+		tracef("setting default Writer as os.Stdout")
+		cmd.Writer = os.Stdout
+	}
+
+	if cmd.ErrWriter == nil {
+		tracef("setting default ErrWriter as os.Stderr")
+		cmd.ErrWriter = os.Stderr
+	}
+
+	if cmd.AllowExtFlags {
+		tracef("visiting all flags given AllowExtFlags=true")
+		// add global flags added by other packages
+		flag.VisitAll(func(f *flag.Flag) {
+			// skip test flags
+			if !strings.HasPrefix(f.Name, ignoreFlagPrefix) {
+				cmd.Flags = append(cmd.Flags, &extFlag{f})
+			}
+		})
+	}
+
+	for _, subCmd := range cmd.Commands {
+		tracef("setting sub-command parent as self")
+		subCmd.parent = cmd
+
+		tracef("assigning subCmd.flagCategories from subCmd.Flags")
+		subCmd.flagCategories = newFlagCategoriesFromFlags(subCmd.Flags)
```

Comment:
Does this need to be done here and not in cmd.Setup ?

Reply:
oooh yeah this looks like a duplicate

thank you!